internal/lang/eval.go

```diff
@@ -343,196 +340,253 @@ func (s *Scope) evalContext(parent *hcl.EvalContext, refs []*addrs.Reference, se
+		rawSubj = addr.ContainingResource()
+	case addrs.ModuleCallInstance:
+		rawSubj = addr.Call
+	case addrs.ModuleCallInstanceOutput:
+		rawSubj = addr.Call.Call
+	}
 
-		case addrs.ForEachAttr:
-			val, valDiags := normalizeRefValue(s.Data.GetForEachAttr(subj, rng))
-			diags = diags.Append(valDiags)
-			forEachAttrs[subj.Name] = val
+	normalizeRefValue := func(val cty.Value, inDiags tfdiags.Diagnostics) cty.Value {
+		normV, normDiags := normalizeRefValue(val, inDiags)
+		diags = diags.Append(normDiags)
+		return normV
+	}
 
-		case addrs.OutputValue:
-			val, valDiags := normalizeRefValue(s.Data.GetOutput(subj, rng))
-			diags = diags.Append(valDiags)
-			outputValues[subj.Name] = val
+	switch subj := rawSubj.(type) {
+	case addrs.Resource:
+		diags = diags.Append(b.putResourceValue(subj, rng))
 
-		case addrs.Check:
-			val, valDiags := normalizeRefValue(s.Data.GetCheckBlock(subj, rng))
-			diags = diags.Append(valDiags)
-			outputValues[subj.Name] = val
-		case addrs.ProviderFunction:
-			// Inject function directly into context
-			if _, ok := ctx.Functions[subj.String()]; !ok {
-				fn, fnDiags := s.ProviderFunctions(subj, rng)
-				diags = diags.Append(fnDiags)
+	case addrs.ModuleCall:
+		b.wholeModules[subj.Name] = normalizeRefValue(b.s.Data.GetModule(subj, rng))
 
-				if !fnDiags.HasErrors() {
-					ctx.Functions[subj.String()] = *fn
-				}
-			}
-		default:
-			// Should never happen
-			panic(fmt.Errorf("Scope.buildEvalContext cannot handle address type %T", rawSubj))
-		}
+	case addrs.InputVariable:
+		b.inputVariables[subj.Name] = normalizeRefValue(b.s.Data.GetInputVariable(subj, rng))
+
+	case addrs.LocalValue:
+		b.localValues[subj.Name] = normalizeRefValue(b.s.Data.GetLocalValue(subj, rng))
+
+	case addrs.PathAttr:
+		b.pathAttrs[subj.Name] = normalizeRefValue(b.s.Data.GetPathAttr(subj, rng))
+
+	case addrs.TerraformAttr:
+		b.terraformAttrs[subj.Name] = normalizeRefValue(b.s.Data.GetTerraformAttr(subj, rng))
+
+	case addrs.CountAttr:
+		b.countAttrs[subj.Name] = normalizeRefValue(b.s.Data.GetCountAttr(subj, rng))
+
+	case addrs.ForEachAttr:
+		b.forEachAttrs[subj.Name] = normalizeRefValue(b.s.Data.GetForEachAttr(subj, rng))
+
+	case addrs.OutputValue:
+		b.outputValues[subj.Name] = normalizeRefValue(b.s.Data.GetOutput(subj, rng))
+
+	case addrs.Check:
+		b.outputValues[subj.Name] = normalizeRefValue(b.s.Data.GetCheckBlock(subj, rng))
+
+	default:
+		// Should never happen
+		panic(fmt.Errorf("Scope.buildEvalContext cannot handle address type %T", rawSubj))
+	}
+
+	return diags
+}
+
+func (b *evalVarBuilder) putResourceValue(res addrs.Resource, rng tfdiags.SourceRange) tfdiags.Diagnostics {
+	var into map[string]map[string]cty.Value
+
+	//nolint:exhaustive // InvalidResourceMode is checked in default.
+	switch res.Mode {
+	case addrs.ManagedResourceMode:
+		into = b.managedResources
+	case addrs.DataResourceMode:
+		into = b.dataResources
+	default:
+		panic(fmt.Errorf("unsupported ResourceMode %s", res.Mode))
 	}
```

Comment:
nit - I think we prefer to include `InvalidResourceMode` here, instead of having a `nolint:exhaustive` here

```suggestion
	switch res.Mode {
	case addrs.ManagedResourceMode:
		into = b.managedResources
	case addrs.DataResourceMode:
		into = b.dataResources
	case addrs.InvalidResourceMode:
		fallthrough
	default:
		panic(fmt.Errorf("unsupported ResourceMode %s", res.Mode))
	}
```